os/drivers/input/touchscreen.c

```diff
@@ -0,0 +1,358 @@
+ ****************************************************************************/
+
+static const struct file_operations g_touchdev_fileops = {
+	touch_open,		/* open */
+	touch_close,	/* close */
+	touch_read,		/* read */
+	touch_write,	/* write */
+	NULL,			/* seek */
+	NULL,			/*ioctl */
+	touch_poll,
+};
+
+/************************************************************************************
+ * Name: touch_semtake
+ ************************************************************************************/
+
+static inline int touch_semtake(FAR sem_t *sem, bool errout)
+{
+	/* Take a count from the semaphore, possibly waiting */
+	if (sem_wait(sem) < 0) {
+		/* EINTR is the only error that we expect */
+		int errcode = get_errno();
+		DEBUGASSERT(errcode == EINTR);
+		if (errout) {
+			return errcode;
+		}
+	}
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_semgive
+ ****************************************************************************/
+
+static inline void touch_semgive(sem_t *sem)
+{
+	sem_post(sem);
+}
+
+/****************************************************************************
+ * Name: touch_open
+ ****************************************************************************/
+
+static int touch_open(FAR struct file *filep)
+{
+	FAR struct touchscreen_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	if (priv->crefs == 0) {
+		priv->ops->touch_enable(priv);
+	}
+	priv->crefs++;
+	DEBUGASSERT(priv->crefs > 0);
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_close
+ ****************************************************************************/
+
+static int touch_close(FAR struct file *filep)
+{
+	FAR struct touchscreen_s *priv;
+	priv = filep->f_inode->i_private;
+
+	if (!priv) {
+		return -EINVAL;
+	}
+
+	touch_semtake(&priv->sem, false);
+	DEBUGASSERT(priv->crefs > 0);
+	priv->crefs--;
+	if (priv->crefs == 0) {
+		priv->ops->touch_disable(priv);
+#ifndef CONFIG_DISABLE_POLL
+		/* Check if this file is registered in a list of waiters for polling.
+		* For example, when task A is blocked by calling poll and task B try to terminate task A,
+		* a pollfd of A remains in this list. If it is, it should be cleared.
+		*/
+		(void)touch_semtake(&priv->pollsem, false);
+		for (int i = 0; i < CONFIG_TOUCH_NPOLLWAITERS; i++) {
+			struct pollfd *fds = priv->fds[i];
+			if (fds && (FAR struct file *)fds->filep == filep) {
+				priv->fds[i] = NULL;
+			}
+		}
+		touch_semgive(&priv->pollsem);
+#endif
+	}
+	touch_semgive(&priv->sem);
+	return OK;
+}
+
+/****************************************************************************
+ * Name: touch_read
+ ****************************************************************************/
+
+static ssize_t touch_read(FAR struct file *filep, FAR char *buffer, size_t buflen)
```

Comment:
as per my previous reveiew, it should be moved to ioctl.
Not all device just return touched data, some other process before collect touch data can be required.
Please do not use read/write to collect or send data, if possible. It's not a good way.

Reply:
I think touch_read and touch_write are fine because all devices require those. 
We can add a separate ioctl which can be used to set configuration/setting before calling touch_read/touch_write in future devices.
@ewoodev should we add a separate ioctl?

Reply:
No, some arguments or pre process can be added. That's why most of character device provide ioctl to send or receive data from application.
Also Tizen Lite follow posix, in POSIX, character device return private data on iNode when read requested.